### PR TITLE
Cqlpg 89 sort by id should directly use primary key 

### PR DESCRIPTION
**Pull request by evaluk**
fairly simple and no frills direction of the sort by field to be the index if it matches the pkColumnName.  Added one unit test to test it out

### CODE REVIEW

**Review by julianladisch — CHANGES_REQUESTED** [review]
Please fix failing unit test pKeyDefault.

Please use the https://dev.folio.org/guidelines/contributing/#coding-style namely https://www.oracle.com/technetwork/java/javase/documentation/codeconventions-141388.html , you may want to configure your IDE/editor to use cql2pgjson-java's .editorconfig file as explained in the first link.

File: cql2pgjson/src/main/java/org/z3950/zing/cql/cql2pgjson/CQL2PgJSON.java [hunk 1]
```diff
@@ -472,9 +472,24 @@ private SqlSelect toSql(CQLSortNode node) throws QueryValidationException {
...
         desc = " DESC";
       }  // ASC not needed, it's Postgres' default
-
+      
       IndexTextAndJsonValues vals = getIndexTextAndJsonValues(modifierSet.getBase());
-
+      String sortbyClause  = wrapInLowerUnaccent(vals.indexText);
+      //determine if sort is primary key 
+      if(dbTable != null) {
+        String pkColumnName = dbTable.optString("pkColumnName", /* default = */ "id");
```
Comment (julianladisch) [inline cql2pgjson/src/main/java/org/z3950/zing/cql/cql2pgjson/CQL2PgJSON.java]: Can you move both instances of this duplicated code into a method, for example getPkColumnName() that returns the default if dbTable is null?

**Review by julianladisch — CHANGES_REQUESTED** [review]
_id still doesn't work, see new unit test.

What do you think about https://github.com/folio-org/cql2pgjson-java/compare/CQLPG-89-sortBy-id-primary-key ?

File: cql2pgjson/src/main/java/org/z3950/zing/cql/cql2pgjson/CQL2PgJSON.java [hunk 2]
```diff
@@ -472,30 +472,54 @@ private SqlSelect toSql(CQLSortNode node) throws QueryValidationException {
...
       // We assume that a CREATE INDEX for this has been installed.
-      order.append(wrapInLowerUnaccent(vals.indexText)).append(desc);
+      order.append(sortbyClauseText).append(desc);
     }
     return new SqlSelect(where, order.toString());
   }
 
+  private String getPkColumnName() {
+    String pkColumnName = dbTable.optString("pkColumnName", /* default = */ "id");
+    return pkColumnName;
```
Comment (julianladisch) [inline cql2pgjson/src/main/java/org/z3950/zing/cql/cql2pgjson/CQL2PgJSON.java]: Please check for dbTable==null. If it is null return the default "id".

File: cql2pgjson/src/main/java/org/z3950/zing/cql/cql2pgjson/CQL2PgJSON.java [hunk 3]
```diff
@@ -472,30 +472,54 @@ private SqlSelect toSql(CQLSortNode node) throws QueryValidationException {
...
         desc = " DESC";
       }  // ASC not needed, it's Postgres' default
-
       IndexTextAndJsonValues vals = getIndexTextAndJsonValues(modifierSet.getBase());
-
+      //prepare sort by clause as if the sort by is not the primary key
+      String sortbyClauseText  = wrapInLowerUnaccent(vals.indexText);
+      String sortByClauseIndex = vals.indexJson;
+      //determine if sort is primary key 
+      if(dbTable != null) {
```
Comment (julianladisch) [inline cql2pgjson/src/main/java/org/z3950/zing/cql/cql2pgjson/CQL2PgJSON.java]: This null check should go into the getPkColumnName() method.

File: cql2pgjson/src/main/java/org/z3950/zing/cql/cql2pgjson/CQL2PgJSON.java [hunk 4]
```diff
@@ -472,30 +472,54 @@ private SqlSelect toSql(CQLSortNode node) throws QueryValidationException {
...
+        if(actualModifierBase.equals("id")) {
+          actualModifierBase = pkColumnName;
+        }
+        if(pkColumnComparator.equals(actualModifierBase)) {
+          //if it is we short circuit the json version of the column and just use the pkColumnName 
+          sortbyClauseText =  pkColumnName;
+          sortByClauseIndex = pkColumnName;
+        } else {
+          
+        }
```
Comment (julianladisch) [inline cql2pgjson/src/main/java/org/z3950/zing/cql/cql2pgjson/CQL2PgJSON.java]: Please explain what this empty else clause is intended for.

**Review by julianladisch — CHANGES_REQUESTED** [review]
Lines 481-495 are 15 lines, create three variables, make two if tests and have three assignments.
An equivalent solution with only 4 lines has been shown in https://github.com/folio-org/cql2pgjson-java/compare/CQLPG-89-sortBy-id-primary-key (note that the unit tests are exactly the same).

Please explain what the advantage of the 15 lines solution over the 4 lines solutions are.

File: cql2pgjson/src/main/java/org/z3950/zing/cql/cql2pgjson/CQL2PgJSON.java (hunk 4 above)
Comment (evaluk) [reply cql2pgjson/src/main/java/org/z3950/zing/cql/cql2pgjson/CQL2PgJSON.java]: resolved with a comment in code

File: cql2pgjson/src/main/java/org/z3950/zing/cql/cql2pgjson/CQL2PgJSON.java (hunk 3 above)
Comment (evaluk) [reply cql2pgjson/src/main/java/org/z3950/zing/cql/cql2pgjson/CQL2PgJSON.java]: moved this line to the function and returned null if found to be null

Comment (julianladisch) [reply cql2pgjson/src/main/java/org/z3950/zing/cql/cql2pgjson/CQL2PgJSON.java]: Thre default pkColumnName is `id`, therefore you should return `id` if dbTable is null.

File: cql2pgjson/src/main/java/org/z3950/zing/cql/cql2pgjson/CQL2PgJSON.java [hunk 5]
```diff
@@ -472,30 +472,56 @@ private SqlSelect toSql(CQLSortNode node) throws QueryValidationException {
...
+      String actualModifierBase = modifierSet.getBase();
+      //_id is actually referred to as id by cql so we need to check for the known cql value
+
+      if(actualModifierBase.equals("id")) {
+        actualModifierBase = pkColumnName;
+      }
+      if(pkColumnComparator != null && pkColumnComparator.equals(actualModifierBase)) {
+        //if it is we short circuit the json version of the column and just use the pkColumnName 
+        sortbyClauseText =  pkColumnName;
+        sortByClauseIndex = pkColumnName;
```
Comment (julianladisch) [inline cql2pgjson/src/main/java/org/z3950/zing/cql/cql2pgjson/CQL2PgJSON.java]: If we know that we use pkColumnName there is no need to continue with testing NUMBER, "number" and "integer". Instead the pkColumnName should be appended to order and `continue` should be used to immediately skip to the next element of the for loop. Similar to the CqlSortType.NUMBER if-clause.

File: cql2pgjson/src/main/java/org/z3950/zing/cql/cql2pgjson/CQL2PgJSON.java [hunk 6]
```diff
@@ -472,30 +472,56 @@ private SqlSelect toSql(CQLSortNode node) throws QueryValidationException {
...
+      
+      String pkColumnName = getPkColumnName();
+      String pkColumnComparator = pkColumnName;
+      String actualModifierBase = modifierSet.getBase();
+      //_id is actually referred to as id by cql so we need to check for the known cql value
+
+      if(actualModifierBase.equals("id")) {
+        actualModifierBase = pkColumnName;
+      }
+      if(pkColumnComparator != null && pkColumnComparator.equals(actualModifierBase)) {
```
Comment (julianladisch) [inline cql2pgjson/src/main/java/org/z3950/zing/cql/cql2pgjson/CQL2PgJSON.java]: pkColumnComparator is never null.

Comment (julianladisch) [inline cql2pgjson/src/main/java/org/z3950/zing/cql/cql2pgjson/CQL2PgJSON.java]: These two if-clauses can be replaced by this:

      if (modifierSet.getBase().equals("id")) {
        order.append(getPkColumnName()).append(desc);
        continue;
      }

Comment (evaluk) [reply cql2pgjson/src/main/java/org/z3950/zing/cql/cql2pgjson/CQL2PgJSON.java]: ok, yes true but I made it be null in a commit, and had to check this.  Trying to handle feedback as fast as possible and we are starting to step on each other at this point.

**Review by julianladisch — APPROVED** [review]

**Pull request merged**